rtengine/demosaic_algos.cc

```diff
@@ -3554,70 +3550,90 @@ void RawImageSource::dcb_pp(float (*image)[4], int x0, int y0)
-            float current = 4.f * image[indx][3] +
-                            2.f * (image[indx + u][3] + image[indx - u][3] + image[indx + 1][3] + image[indx - 1][3])
-                            + image[indx + v][3] + image[indx - v][3] + image[indx - 2][3] + image[indx + 2][3];
 
-            f[0] = (float)(image[indx - u][1] + image[indx + u][1]) / (2.f + 2.f * image[indx][c]);
-            f[1] = 2.f * image[indx - u][1] / (2 + image[indx - v][c] + image[indx][c]);
-            f[2] = (float)(image[indx - u][1] + image[indx - w][1]) / (2.f + 2.f * image[indx - v][c]);
-            f[3] = 2.f * image[indx + u][1] / (2 + image[indx + v][c] + image[indx][c]);
-            f[4] = (float)(image[indx + u][1] + image[indx + w][1]) / (2.f + 2.f * image[indx + v][c]);
+            float current = 4 * map[indx] +
+                            2 * (map[indx + u] + map[indx - u] + map[indx + 1] + map[indx - 1])
+                            + map[indx + v] + map[indx - v] + map[indx - 2] + map[indx + 2];
+
+            float currPix = image[indx][c];
 
-            g1 = (f[0] + f[1] + f[2] + f[3] + f[4] - max(f[1], f[2], f[3], f[4]) - min(f[1], f[2], f[3], f[4])) / 3.f;
+            f0 = (float)(image[indx - u][1] + image[indx + u][1]) / (1.f + 2.f * currPix);
+            f1 = 2.f * image[indx - u][1] / (1.f + image[indx - v][c] + currPix);
+            f2 = 2.f * image[indx + u][1] / (1.f + image[indx + v][c] + currPix);
 
-            f[0] = (float)(image[indx - 1][1] + image[indx + 1][1]) / (2.f + 2.f * image[indx][c]);
-            f[1] = 2.f * image[indx - 1][1] / (2 + image[indx - 2][c] + image[indx][c]);
-            f[2] = (float)(image[indx - 1][1] + image[indx - 3][1]) / (2.f + 2.f * image[indx - 2][c]);
-            f[3] = 2.f * image[indx + 1][1] / (2 + image[indx + 2][c] + image[indx][c]);
-            f[4] = (float)(image[indx + 1][1] + image[indx + 3][1]) / (2.f + 2.f * image[indx + 2][c]);
+            g1 = f0 + f1 + f2;
 
-            g2 = (f[0] + f[1] + f[2] + f[3] + f[4] - max(f[1], f[2], f[3], f[4]) - min(f[1], f[2], f[3], f[4])) / 3.f;
+            h0 = (float)(image[indx - 1][1] + image[indx + 1][1]) / (1.f + 2.f * currPix);
+            h1 = 2.f * image[indx - 1][1] / (1.f + image[indx - 2][c] + currPix);
+            h2 = 2.f * image[indx + 1][1] / (1.f + image[indx + 2][c] + currPix);
 
+            g2 = h0 + h1 + h2;
+		
+            // new green value
             assert(indx >= 0 && indx < u * u);
-            image[indx][1] = (2.f + image[indx][c]) * (current * g1 + (16.f - current) * g2) * 0.0625f;
+            currPix *= (current * g1 + (16.f - current) * g2) / 48.f;
+
+            // get rid of the overshot pixels
+            float minVal = min(image[indx - 1][1], min(image[indx + 1][1], min(image[indx - u][1], image[indx + u][1])));
```

Comment:
@heckflosse As to the first point: It's a pity that often the more elegant code is the slower. But we have `min()` and `max()` with up to four parameters in [`rt_math.h`](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/rt_math.h#L82), which could be handy here.

As to the second point: Good catch. Also requires a facepalm when thinking about it.